### PR TITLE
w261567387 Fix Worker Health Checks

### PR DESCRIPTION
**Pull request by whatisdot**
Closes https://www.wrike.com/open.htm?id=261567387

# TL;DR
This creates ECS Services that give us all of the EB functionality we require, while only using one EB environment. Workers now start with zero tasks running by default, and must be scaled up manually. This matches Heroku functionality.

# Problem Statement
EB Worker environments are purpose-built to send POST requests to the application it is hosting. This is ill-suited for all of our worker and background processes. EB health checks shows all of our worker environments as unhealthy until they expose an endpoint for receiving requests. This functionality cannot be disabled or adjusted, and adding core logic as a work-around is additional effort that doesn't directly address the problem.

Multiple elastic beanstalk environments for the same QA/Stage are difficult to manage, as they clutter the view and require deep click-through to see related metrics and logs. 

Endpoints are also being exposed for background processes which are never designed to be directly reachable.

Deploying to multiple elastic beanstalk endpoints also presents problems, as it allows potential differences between environment variables across environments.

# Solution
By leveraging ECS Services for the ECS cluster associated with the 'web' default elastic beanstalk environment, we can avoid the need to create additional EB environments for worker processes.

screen shot 2018-09-12 at 3 43 29 pm

ECS Services offer a similar value-add to Elastic Beanstalk. What follows is a breakdown of each EB configuration area, represented as an ECS service:

## Software (Logging Support)
Elastic beanstalk requires that we navigate to CloudWatch and click on each individual log group. Logging support for Fargate services is automatic, and provides an interface for viewing the logs of all tasks executing for a specific ECS Service.
screen shot 2018-09-12 at 3 46 56 pm

There is also no way to configure environment variables in the 

## (EC2) Instances
...are no longer a concern for background workers. The Fargate type abstracts the way containers are hosted, removing the need for us to monitor EC2 instances.

## Capacity
Each ECS Service within a cluster (for example QA5) can automatically scale independently of other services. It can also be configured with scaling triggers.
screen shot 2018-09-12 at 3 40 27 pm

The interface for doing this is also more helpful, with a recent graph of performance.
screen shot 2018-09-12 at 3 39 34 pm

## Load Balancing
ECS Services can be configured for load balancing, but our current worker processes have no need. This has been disabled by default.

## Rolling Updates & Deployments
Version updates are all rolling by default, and configured by the "healthy percentage" options.
screen shot 2018-09-12 at 3 48 04 pm

All deployments are clearly logged for individual ECS Services:
screen shot 2018-09-12 at 4 04 34 pm

## Security
Security is also more fine-grained, with an assumed role for the "ECS Task" and an assumed role for the "ECS Service". Elastic beanstalk wraps both of these in one assumed role.

SSH-ing into a host is not permitted for Fargate, for added security.

## Monitoring
ECS Services can be monitored from within their cluster, rather than needing to navigate to a separate EB environment.
screen shot 2018-09-12 at 4 12 48 pm

## Health Checks
ECS Services offer more fine-grained health check capabilities. The task definitions can be used to define [custom commands](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definition_healthcheck) to run to determine if a container is healthy.

## Managed Updates
The version of ECS Service platform can be updated by editing an existing service, or by recreating a service. This can also be done through a git-scripts update to the configuration we pass during deployment.
screen shot 2018-09-12 at 4 15 36 pm

## Notifications
...are currently not being used. These should be part of an integrated reporting system, and can also exist in the form of CloudWatch alarms.

## Network
This is configured by subnet, similar to EB. Currently, we use the default subnets for the region. A Public IP address is used so that ECR can be accessed during deployment, but it [has been documented](https://github.com/aws/amazon-ecs-agent/issues/1128#issuecomment-351545461) that this can be avoided with proper subnet configuration.

## Database
There is no default assumptions in ECS, so there are also no unused database configuration pages.

# Adapting
This implementation will synch the ECS Services with every deployment. Developers can add and remove workers, deploy and immediately see the results. Trying to do this with our current setup will break when trying to deploy to QA.

# Testing

Checkout a migrated project and create a throw-away branch off of master.

Edit the Procfile and add an additional line with a different name. Commit your changes (no "git push" is needed).
screen shot 2018-09-12 at 4 30 58 pm

Run `git_push_eb` to an environment. Note the new notifications in the logs, and that we only deploy to a single EB environment:
```bash
[2018-09-12 16:33:01,368]: Updating task definition for: elasticsearch_indexing_worker
[2018-09-12 16:33:04,626]: Updating service for elasticsearch_indexing_worker
[2018-09-12 16:33:05,473]: Updating task definition for: some_other_worker_process
[2018-09-12 16:33:08,656]: Creating service for some_other_worker_process
```

Navigate to the ECS cluster associated with the *-web environment for the QA you have pushed too. Confirm that the service is now present, and has zero tasks running.

screen shot 2018-09-12 at 4 41 37 pm

Edit the new Service and scale it's desired count to 1. Update the service and confirm an additional task has been created in the cluster and it is running as expected:

screen shot 2018-09-12 at 4 43 30 pm

screen shot 2018-09-12 at 4 43 37 pm

screen shot 2018-09-12 at 4 43 48 pm

screen shot 2018-09-12 at 4 46 00 pm

screen shot 2018-09-12 at 4 46 14 pm

screen shot 2018-09-12 at 4 52 17 pm

Now edit the Procfile to remove the additional line. Commit your changes.
screen shot 2018-09-12 at 4 47 53 pm

Do another push to the same environment. The logs should show that the service cannot be removed with a running task.
```bash
[2018-09-12 16:53:42,478]: Removing service no longer referenced: some_other_worker_process
[2018-09-12 16:53:43,384]:
An error occurred (InvalidParameterException) when calling the DeleteService operation: The service cannot be stopped while the primary deployment is scaled above 0.
```

Now adjust the service to set the task count back to zero. Wait for it to finish updating, then try another push. You should see the service is now removed.
```bash
[2018-09-12 16:57:15,316]: Updating task definition for: elasticsearch_indexing_worker
[2018-09-12 16:57:18,408]: Updating service for elasticsearch_indexing_worker
[2018-09-12 16:57:19,220]: Removing service no longer referenced: some_other_worker_process
```

screen shot 2018-09-12 at 4 57 53 pm

### CODE REVIEW

**Comment by whatisdot** [comment]
@santi-h The best way to get rid of bad environment health checks is to get rid of the environment.

This PR replaces the EB environment for workers with an ECS Service counterpart on the same cluster. Please review and ask questions, as this is meant to be a comprehensive change.

**Comment by whatisdot** [comment]
I also disagree with CodeClimate, and thing the size of `PushEB` is appropriate. If you agree with CodeClimate, please provide an explanation of how you would refactor the module to satisfy the 250-line rule.

**Comment by jeet4320** [comment]
as discussed, please add below code to select env on which deployment will be carried out and add logic to increment build number under `message`

```
  def deploy(self):
    self.logger.info("Deploying to all web and worker envs.")
    eb_args = [
      'deploy',
      '--env-group-suffix', hashlib.sha1(self.project_name()).hexdigest()[:10],
      '--message', self.args.environment + '- 2',
      '--region', self.region(),
      '--profile', DEFAULT_PROFILE,
      '--label', self.sha() ,
      EB.environment_name(self.project_name(), self.args.environment, 'web')
    ]
    app = EBCli(argv=eb_args)
    ebrun.run_app(app)
```

**Comment by santi-h** [comment]
Ya don't listen to codeclimate, we should disable the method length check

File: lib/push_eb.py [hunk 1]
```diff
@@ -218,20 +198,22 @@ def update_scheduled_task_configurations(self):
...
-  def create_or_update_definition(self, name, task):
-    self.logger.info("Updating task definitions.")
+  def create_or_update_definition(self, name, command):
+    self.logger.info("Updating task definition for: "+name)
+    env_vars = self.get_environment_variables()
     response = ECS.register_task_definition({
-      'env_vars': self.get_environment_variables('web'),
-      'family': self.cluster_name(name),
+      'command': command,
+      'env_vars': [{'name': key, 'value': env_vars[key]} for key in env_vars],
```
Comment (jeet4320) [inline lib/push_eb.py]: If we add env variable to EB env from AWS UI, then EB update env won't update env variables on ECS tasks. We could create script to add env variable and update task definitions, ask users to update env variables using script and not from UI.

Comment (whatisdot) [reply lib/push_eb.py]: Yes. Using a script to update a single or group of ENV VAR for a given environment would be the best option at the moment. Currently, the best way to adjust an ENV VAR would be to update the value in EB, then do a deployment. Deployments update environment variables for ECS Services every time.

**Comment by whatisdot** [comment]
I've updated the scripts to use the Git SHA for the Version Label, and environment name for Description.
screen shot 2018-09-13 at 11 16 17 am

This has another benefit, because we can now see the Git SHA on the EB environment dashboard page.
screen shot 2018-09-13 at 11 24 39 am

File: lib/run_eb.py [hunk 2]
```diff
@@ -18,10 +18,10 @@ def parse_args(self):
     return args, additional_args
 
   def command(self):
-    current_version = self.get_current_version()
+    current_version = self.sha()
```
Comment (jeet4320) [inline lib/run_eb.py]: we should keep `get_current_version` method and pull `sha` from `describe-application-versions`. Using local `sha` means I have to be on that `sha` to run `git_run_eb`. 

Anyone should be able to run `git_run_eb` from any git branch, and it will pull the latest code on the env

File: lib/push_eb.py (hunk 1 above)
Comment (jeet4320) [reply lib/push_eb.py]: You mean deployment using `git-scripts`? Would deployment from AWS UI update env variables of ECS tasks present under cluster of the EB env?

File: lib/push_eb.py [hunk 3]
```diff
@@ -1,29 +1,28 @@
...
-from aws.ecs import ECS
 from aws.cloudwatch.events import Events
 from aws.cloudwatch.targets import Targets
+from aws.eb import EB
 from aws.ecs import ECS
+from aws.ec2 import EC2
 from aws.procfile import Procfile
 from aws.scheduled_task_configuration import ScheduledTaskConfiguration
-import ebcli.core.ebrun as ebrun
 from ebcli_overrides.core.ebcore import EB as EBCli
```
Comment (jeet4320) [inline lib/push_eb.py]: can we remove `ebcli_overrides` as we are using `eb deploy` and no need to override it. `eb deploy` provides ability to change `application version`, `description`.

Comment (whatisdot) [reply lib/push_eb.py]: I want to limit the scope of this PR, so we can add a cleanup task for this.

**Comment by whatisdot** [comment]
@santi-h @nickstanish requesting review, please

**Comment by whatisdot** [comment]
@nickstanish @jeet4320 I've pushed a branch named `fake-branch-dnm` to the vice-backend project. You can look at the commit log to see how I tested this: editing the lines in the `Procfile` and creating/modifying `scheduled_tasks.yml`.

I ran full feature tests for affected functionality:
* `git_terraform_commands` to generate the scripts to create a QA9 environment.
* `git_copy_env_eb` to copy environment variables from `qa5-web` to `qa9`.
* `git_push_eb qa9` to push changes to the new environment.
* `git_run_eb qa9 rake db:setup` to set up the database.

This update will require documentation that describes how environments have been changed. Fortunately, I've been putting together a Google presentation to describe the features now included in git-scripts for AWS integrations. I will update it to reflect the new behavior. Details will be in our Slack channel.

### Pre-merge tasks
- [ ] Create the new QA environments for all existing projects.
- [ ] Update documentation on how to use the ECS clusters.

### Post-merge tasks
- [ ] Tell everyone to update their git scripts immediately.
- [ ] Distribute documentation on how to use ECS clusters.
- [ ] Swap the IAM environment policies for the associated team groups.

File: lib/copy_env_eb.py [hunk 4]
```diff
@@ -25,7 +25,7 @@ def parse_args(self):
     return [args, None]
 
   def command(self):
-    env_vars = self.get_environment_variables('web')
+    env_vars = self.get_environment_variables()
```
Comment (jeet4320) [inline lib/copy_env_eb.py]: we should remove loop at [here](https://github.com/bodyshopbidsdotcom/git-scripts/pull/382/files#diff-eb8b50a8079711ef2c405d0f7a50e1c8R47) as it is calling env update on same `web` env multiple times

Comment (whatisdot) [reply lib/copy_env_eb.py]: @jeet4320 The loop has been removed. We will still need to do a push to have the environment variables propagate to the worker ECS Services, but we can tackle that as a separate issue. This should cover a majority of cases.
